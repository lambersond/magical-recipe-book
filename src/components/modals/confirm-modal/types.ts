import type { ModalProps } from '../types'

export interface ConfirmModalProps extends ModalProps {
  onConfirm: VoidFunction
  message?: string | React.ReactNode
  title?: string
  confirmButtonText?: string
  cancelButtonText?: string
  hideCancelButton?: boolean
}
