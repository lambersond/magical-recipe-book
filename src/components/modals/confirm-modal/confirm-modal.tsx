import { Modal } from '@/components/common'
import { useModals } from '@/hooks/use-modals'
import { primaryButton, secondaryButton } from '@/utils/styles'
import type { ConfirmModalProps } from './types'

export function ConfirmModal({
  onConfirm,
  open = true,
  message = 'Are you sure you want to proceed?',
  title = 'Confirmation',
}: Readonly<ConfirmModalProps>) {
  const { closeModal } = useModals()

  const onClose = () => {
    closeModal('ConfirmModal')
  }

  const handleConfirm = () => {
    onConfirm()
    onClose()
  }

  return (
    <Modal title={title} isOpen={open} onClose={onClose}>
      <p className='my-8 font-bold italic'>{message}</p>
      <div className='flex justify-end mt-4 gap-2'>
        <button onClick={onClose} className={secondaryButton}>
          Cancel
        </button>
        <button onClick={handleConfirm} className={primaryButton}>
          Confirm
        </button>
      </div>
    </Modal>
  )
}
