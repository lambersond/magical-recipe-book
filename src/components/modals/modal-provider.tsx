import {
  type Dispatch,
  type ReactNode,
  createContext,
  useReducer,
  Suspense,
  use,
} from 'react'
import { MODALS } from '@/modals'
import type { ModalMap, Modal, ModalState, ModalAction } from '@/types/modals'

export const ModalStateCtx = createContext<ModalState | undefined>(undefined)
export const ModalDispatchCtx = createContext<
  Dispatch<ModalAction<keyof ModalMap>> | undefined
>(undefined)

const reducer = (state: ModalState, action: ModalAction<Modal>): ModalState => {
  switch (action.type) {
    case 'open': {
      return { ...state, [action.modal]: { open: true, props: action.props } }
    }
    case 'close': {
      return { ...state, [action.modal]: { open: false, props: {} } }
    }
    default: {
      throw new Error(`Unhandled action type: ${action}`)
    }
  }
}

function ModalManager() {
  const state = use(ModalStateCtx)

  if (!state) return

  const modals = Object.entries(state).map(([modal, state]) => {
    const Component = MODALS[modal as Modal]
    const props = state.props as any

    return (
      <Suspense key={modal}>
        <Component open={state.open} {...props} />
      </Suspense>
    )
  })

  return <>{modals}</>
}

export function ModalProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(reducer, {} as ModalState)

  return (
    <ModalDispatchCtx value={dispatch}>
      <ModalStateCtx value={state}>
        {children}
        <ModalManager />
      </ModalStateCtx>
    </ModalDispatchCtx>
  )
}
