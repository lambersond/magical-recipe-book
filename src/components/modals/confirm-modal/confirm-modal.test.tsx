import { renderWithModalProvider, useUser } from '@test-utils'
import { ModalProvider } from '../modal-provider'
import { ConfirmModal } from './confirm-modal'

const mockCloseModal = jest.fn()

jest.mock('../../../hooks/use-modals', () => ({
  useModals: () => ({
    closeModal: mockCloseModal,
  }),
}))

describe('compoenents/modals/confirm-modal', () => {
  it('should render defaults', () => {
    expect(
      renderWithModalProvider(
        <ConfirmModal open onConfirm={jest.fn} />,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should render with passed props', () => {
    expect(
      renderWithModalProvider(
        <ConfirmModal
          open
          title='Test title'
          message='Test message'
          onConfirm={jest.fn}
        />,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should call onConfirm when confirm button is clicked', async () => {
    const onConfirm = jest.fn()

    const { getByText, user } = useUser(
      <ConfirmModal open onConfirm={onConfirm} />,
      { wrapper: ModalProvider },
    )

    await user.click(getByText('Confirm'))

    expect(onConfirm).toHaveBeenCalled()
  })

  it('should call closeModal when cancel button is clicked', async () => {
    const { getByText, user } = useUser(
      <ConfirmModal open onConfirm={jest.fn} />,
      { wrapper: ModalProvider },
    )

    await user.click(getByText('Cancel'))
    expect(mockCloseModal).toHaveBeenCalled()
  })
})
