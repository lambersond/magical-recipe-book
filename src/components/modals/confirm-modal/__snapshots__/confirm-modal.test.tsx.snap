// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`compoenents/modals/confirm-modal should render defaults 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex items-center justify-center z-1200"
  >
    <summary
      class="absolute inset-0 bg-black opacity-75"
      data-testid="modal__backdrop"
    />
    <div
      class="relative bg-paper max-w-2xl w-sm md:w-md z-1000 overflow-hidden shadow-lg h-full min-w-full sm:min-w-sm  sm:rounded-xl sm:h-fit"
      data-testid="modal"
    >
      <div
        class="flex justify-between items-center p-3"
      >
        <p
          class="text-2xl font-bold"
        >
          Confirmation
        </p>
        <button
          class="bg-transparent text-text-primary hover:text-text-primary hover:bg-text-secondary/30 cursor-pointer hover:text-text-primary size-10 border-transparent min-w-fit rounded-full p-1 flex items-center justify-center flex justify-between focus-within:outline-none focus:outline-none"
        >
          <svg
            class="min-w-fit min-h-fit"
            data-testid="CloseIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="overflow-y-auto p-3"
        style="max-height: calc(100vh - 0px);"
      >
        <span
          class="my-8 font-bold italic"
        >
          Are you sure you want to proceed?
        </span>
        <div
          class="flex justify-end mt-4 gap-2"
        >
          <button
            class="px-4 py-2 text-text-secondary border-2 border-border  hover:text-text-primary/70 hover:border-text-primary/70 hover:text-text-primary/70 cursor-pointer rounded-lg text-lg font-bold uppercase"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 bg-error/20 text-error hover:bg-red-800 hover:text-red-300 cursor-pointer rounded-lg text-lg font-bold uppercase"
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`compoenents/modals/confirm-modal should render with passed props 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex items-center justify-center z-1200"
  >
    <summary
      class="absolute inset-0 bg-black opacity-75"
      data-testid="modal__backdrop"
    />
    <div
      class="relative bg-paper max-w-2xl w-sm md:w-md z-1000 overflow-hidden shadow-lg h-full min-w-full sm:min-w-sm  sm:rounded-xl sm:h-fit"
      data-testid="modal"
    >
      <div
        class="flex justify-between items-center p-3"
      >
        <p
          class="text-2xl font-bold"
        >
          Test title
        </p>
        <button
          class="bg-transparent text-text-primary hover:text-text-primary hover:bg-text-secondary/30 cursor-pointer hover:text-text-primary size-10 border-transparent min-w-fit rounded-full p-1 flex items-center justify-center flex justify-between focus-within:outline-none focus:outline-none"
        >
          <svg
            class="min-w-fit min-h-fit"
            data-testid="CloseIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="overflow-y-auto p-3"
        style="max-height: calc(100vh - 0px);"
      >
        <span
          class="my-8 font-bold italic"
        >
          Test message
        </span>
        <div
          class="flex justify-end mt-4 gap-2"
        >
          <button
            class="px-4 py-2 text-text-secondary border-2 border-border  hover:text-text-primary/70 hover:border-text-primary/70 hover:text-text-primary/70 cursor-pointer rounded-lg text-lg font-bold uppercase"
          >
            Cancel
          </button>
          <button
            class="px-4 py-2 bg-error/20 text-error hover:bg-red-800 hover:text-red-300 cursor-pointer rounded-lg text-lg font-bold uppercase"
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
